osdep/w32_register: add register/unregister options for Windows

kasper93 opened the pull request:
I deleted the branch by accident and it nuked the PR, which cannot be reopened for recreated branches. Old PR: #15904

kasper93 commented:
Open questions:

* Should ``--install`` be renamed to ``--register``?

I think users are more familiar with install word. There were voices that install has to move binary somewhere else, but looking at definition https://www.merriam-webster.com/dictionary/install 

> to set up for use or service

it is exactly what it does.

* Should ``--install`` in fact move mpv.exe into user programs?
I prefer keeping it portable, the `--install` / `--uninstall` command are not tied to installation itself. For example you can move binary, install again and it will work. There is no need to uninstall anything, before moving.

* Should archive-exts list be added to mpv's supported formats
I don't see a reason why not. It is still user decision to select application to use. This is handled externally by Windows. Although I agree, mpv it likely not primary choice for archives, it's nice to have it suggested in "open with" drop-down.

Anything else?

--- 

> Most softwares' installer, main exec and uninstaller are detached. And not all sw use the same dir for main exec and uninstaller.

Yes we don't have dedicated installer, which drops dozen of files everywhere. My objective is to make this as clean and light as possible, while still making things work in Windows.

> Some installers could detect the installed one and could even directly uninstall it.

The same as mpv, you can do `mpv --uninstall` from any mpv binary and it will work correctly. In fact while `--install` is handy to be in mpv, because it can be parametrized easily for used extensions, without creating whole another application or using installer for that. The `--uninstall` command is mostly portable. Though I still don't like the idea of dropping files in random places. But this is one of the questions above. If people want this, we can copy the .exe to programs...

Though I see no real issue, if someone uses mpv.exe and removes it, nothing bad happens... Worse case thay have stale entry in control panel, which will be pruned by windows. We can also completely skip uninstall entry in control panel if this is the main concern. But I still don't see why if user breaks it, it's mpv's fault.

> mpv is the only special case I've meet that it could do three jobs by the same single file.

and if it's two files and you remove `uninstall.bat` what happens? mpv.exe is single binary, we really don't need full installer to handle it. Also mpc-hc, foobar2000, lav filters, 7zip all have in place installation. foobar2000 even does it on every start of application. madvr/lav filters have bat files, that lives along .dll, but it's still calls in the .dll regsvr. So if you remove them, well you no longer can uninstall.

avih commented:
Docs changes LGTM. Thanks.

I don't feel strongly about any of the open questions.

Moving the mpv binary would indeed make it a normal installer, but it's more work. I'd imagine there's a good reason that usually an installer application is used (inno setup etc), because it looks simple but I imagine there are many details to take care of - which personally I'm not familiar with.

I think registering mpv inplace is enough. Adding an actual installer would be fine too, but I don't think it's required.

Re the option name, I think the fact that it adds an uninstall entry at the control panel (which is a good thing) kind of makes us choose "install", for consistency. But "register" also a valid choice, because it is indeed a better description of the action.

Bottom line, either one would be OK IMO.

Re icon, not necessary IMO but it would be nicer to have one. I don't think mpv icon would be too bad, but also, there are enough free icons out there, and a generic cog icon or some such would do fine IMO. Or like the original PR did, pick an icon from an existing system file. Just ensure/test that it still works (does the registration) even if the file is missing.

hooke007 commented:
I vote for --register. Not a serious reason: Making it looks more different with mpv-install.bat.

kasper93 commented:
> I vote for --register.

Ok, we can go with `--register` and `--unregister`, while `install`/`uninstall` term might be more familiar to regular joe when manipulation software. Also note I won't use `--deregister`, contrary to what might seem more fighting according to https://english.stackexchange.com/questions/25931/unregister-vs-deregister, but the `--unregister` feels nicer and is what also Windows itself used for operations like that. 

(software like madvr / lav filters, were for years using `install`, hance why this was my first choice, but no matter)

> Moving the mpv binary would indeed make it a normal installer, but it's more work. I'd imagine there's a good reason that usually an installer application is used (inno setup etc), because it looks simple but I imagine there are many details to take care of - which personally I'm not familiar with.

After consideration this is not a job for this command. It looks simple if you forget that normally you have more than single executable to package. Creating installer is still possible and valid on top of this registration command. Note that it can be self-expanding archive, which would serve similar purpose. The point is if you have to handle more files, like additional .dlls, icons, whatever. It makes only sense to have separate packaging solution.

But we need to distinguish what this PR does, which is registration of media plater from packaging the mpv into self-extracting or self-installing single-click solution.

Also like I was hinting on IRC, once you start moving things, you end up looking like malware https://devblogs.microsoft.com/oldnewthing/20230911-00/?p=108749 :) That's of course a joke and not applicable to mpv, but if you want to uninstall the uninstall.exe itself, it's always funny exercise.

Bonus chatter: If we were using cmake, we would have all sort of packaging solutions for free with cpack that integrates with install command. Meson doesn't have this convenience.

> Re icon, not necessary IMO but it would be nicer to have one. I don't think mpv icon would be too bad, but also, there are enough free icons out there, and a generic cog icon or some such would do fine IMO. Or like the original PR did, pick an icon from an existing system file. Just ensure/test that it still works (does the registration/whatever) even if the file is missing.

This is in fact detail. I went back and forth few times, changing this install/uninstall helpers from showing command line window and not. And I settled to show it, which means, it won't have manifest or icon. But that's fine... it should little tiny command line icon.

kasper93 commented:
> I vote for --register. Not a serious reason: Making it looks more different with mpv-install.bat.

Changed everything to `register`. If feels awkward at places, but I will get used to it.

> Perhaps the documentation should be updated to indicate that installation is required to make STMC show the icon and program name.

Added.

> Please consider changing the first sentence to something along those lines, if you too think it can make it clearer.

Added.

--- 

Any more suggestions? I'm not sure what would be useful to add/change here.

avih commented:
> > Please consider changing the first sentence to something along those lines, if you too think it can make it clearer.
> 
> Added.

Thanks.

Apologies for wasting both out time due to me not reading the whole section carefully before embarking on a discussion.

kasper93 commented:
> Is there any reason not to add it to 0.40.0? We hope to replace the old bat installer in shinchiro builds with this soon.

I don't see the reason to rush this through. Windows users doesn't care about stable releases anyway and even if, they can wait.

CharlesMengCA commented:
Can we move these functions to mpv.com?

IMHO, mpv.exe should be a pure player, not installer or uninstaller.

CharlesMengCA commented:
Then maybe we should create another utility exe for install/uninstall or other staff.

Put install/uninstall in the player is a waste. the install/uninstall function will be rarely used, once a year or never.

But the bytes will be loaded every time when we play a thing.

kasper93 commented:
> Putting install/uninstall in the player is a waste.

Waste of what? Bytes? You have over 100MiB static linked binary and worrying about maybe few hundreds bytes of code? (I'm not even sure it is over 100...) Give me a break.

kasper93 commented:
> Pretty soon you'll need a shit NSIS to manage your installation.

This was one of the proposal, or inno setup. I personally want to avoid. The main design decision was to make this portable and light. Hence why it is embedded in mpv and can be used whenever needed. Not only if you have separate script/binary at hand.

Of course proper installer would resolve concerns about removing binary, bloat in binary or anything like that. But it no longer makes it "portable". Currently we can easily switch between different mpv binaries, just bu installing again. There is no artifact left of disc or bloat. Of course there are registry keys and one shortcut, but there is not some uninstallers and other things dropped.

I like the current system, I'm open for proposals how to improve that. But I want to keep it invisible as much as possible. If someone wants to install they can at any time, if not they can just never do it.

CharlesMengCA commented:
After disabled those encoders, my mpv.exe is 51MB although I used libmpv-2.dll most of my time.

I used .NET Framework to write the file association things for both mpv.exe and foobar.exe, hundred lines of code maybe. at lease it is easy to maintain. :)

verygoodlee commented:
`--register-rpath` does not prepend to PATH if mpv launched from terminal.

- `mpv --register --register-rpath=D:\bin\`
- ~~/script/test.lua
```lua
local utils = require('mp.utils')
for _, env in ipairs(utils.get_env_list()) do
    if env:upper():sub(1, 5) == 'PATH=' then
        print(env)
        break
    end
end
```
-  launch mpv via double click mpv.exe and then open console, `D:\bin\` is in PATH.
![1742972527472](https://github.com/user-attachments/assets/f3c02367-d666-4116-8ace-42017b2b60c8)
-  launch mpv from terminal, `D:\bin\` is not in PATH.
![1742972433526](https://github.com/user-attachments/assets/18944903-807e-48be-b26a-4a35302f4a63)

verygoodlee commented:
> > `--register-rpath` does not prepend to PATH if mpv launched from terminal.
> 
> Yes, it does not. Set you environment correctly. PATH is only prepended when run from Windows shell.

mpv may be launched by other applications, such as [umpv](https://github.com/mpv-player/mpv/blob/master/TOOLS/umpv), the `--register-rpath` option doesn't work in these cases, so it does not make much sense, different launch modes result in different behaviors, which can be very confusing for users.

kasper93 commented:
> > > `--register-rpath` does not prepend to PATH if mpv launched from terminal.
> > 
> > 
> > Yes, it does not. Set you environment correctly. PATH is only prepended when run from Windows shell.
> 
> mpv may be launched by other applications, such as [umpv](https://github.com/mpv-player/mpv/blob/master/TOOLS/umpv), the `--register-rpath` option doesn't work in these cases, so it does not make much sense, different launch modes result in different behaviors, which can be very confusing for users.

User inability to understand the feature doesn't render such feature useless. No one is forcing you to use this, if it doesn't fit your workflow.

I've added workaround to documentation. If you run mpv with Start-Process command it will inject PATH. You can create `mpv.ps1` script and other script like `umpv` will use it.

--- 

I've changed to using .bat scripts as per requests. The rest of the patch is mostly unchanged from the last time, I will merge it later, because it looks like this is one of the missing features in our CI builds.

kasper93 commented on meson.build:
```diff
@@ -1813,11 +1814,15 @@ if get_option('cplayer')
-                   override_options: ['b_sanitize=none'],
-                   name_suffix: 'com', install: true)
+        wrapper_sources = 'osdep/win32-console-wrapper.c'
+        if get_option('win32-subsystem') != 'console'
```
Fixed, thanks.

candrapersada commented:
is there an `--unregister-rpath` too?

kasper93 commented:
> is there an `--unregister-rpath` too?
No it's part of register. And cleared on unregister too.

kasper93 commented:
@Obegg `--register` doesn't forcefully change your default applications. It registers mpv as a application that can open `.mp4`. Once you open it once and select mpv as default application, both icon and any next open will use mpv by default.

hooke007 commented:
mpv install never did that. That's must be done by yourself.
